### ISSUE TITLE
Add dark mode

### ISSUE THREAD

**Issue by manuales** [vuejs/vuepress#1865]
#### Are you willing to work on this yourself?

Integration of Dark mode in Vuepress

**Comment by f3ltron** [comment]
That's definitely a cool feature needed maybe it should be specific to the default theme plugin.

Can you work on it ? Do you need help some on boarding etc ?

Feel free to ask

**Comment by vinayakkulkarni** [comment]
just saw @f3ltron 's comment, I've managed to add a toggle button in the NavBar - reference, (https://github.com/vinayakkulkarni/vuepress/tree/feat/dark-theme), need some help regarding stylus as I've never used it before.

**Comment by tolking** [comment]
If create a plugin, you may need to add `!important` in the styles.
If the default theme can support [prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme), it is best.
Or you need to create a new theme like [vuepress-theme-default-prefers-color-scheme](https://github.com/tolking/vuepress-theme-default-prefers-color-scheme)

**Comment by vinayakkulkarni** [comment]
> If create a plugin, you may need to add !important in the styles.

Not really required, you can (should) use css variables `var()` to extend the theme imo

**Comment by songololo** [comment]
I think some sort of more comprehensive distillation of theming in VuePress would be really useful so that it is easier to override with custom theming (without a whole bunch of `!important` statements!)

css variables sound like a step in the right direction.

**Comment by Mister-Hope** [comment]
Wow, I just saw this, I built my own blog with this feature.
My blog https://mrhope.site
My source code: https://github.com/Mister-Hope/blog

I think it is not hard to make a plugin.
If you need any help, just call me.

**Comment by Mister-Hope** [comment]
I did not see any changes on the styles file?
Are you still working on with it? @vinayakkulkarni 
I can sperate my files to release a plugin.
Since the plugin cannot access style files directly, my thought is to inject a style tag to the document when the switch component is mounted(or beforemount).

**Comment by Mister-Hope** [comment]
If you would like to make this feature with the default theme(I think many people would like to use it), then it's easier.

**Comment by vinayakkulkarni** [comment]
@Mister-Hope : Do check this out: https://github.com/vinayakkulkarni/vuepress/tree/feat/dark-theme, I am working on adding the dark theme. Need to integrate the colour palette used by Akruym in dev-tool dark theme.

**Comment by Mister-Hope** [comment]
I checked this, so I am saying I did not see any changes on the style file.

So Akruym is dealing with the style?

**Comment by Mister-Hope** [comment]
![image](https://user-images.githubusercontent.com/33315834/66474898-c9b3bc80-eac4-11e9-92cd-53025c51081a.png)
I think it is not a complex work, since the stylus file in Vuepress is very ordered.
Can't wait to see the feature offical supported

**Comment by Mister-Hope** [comment]
@vinayakkulkarni Could you check [this stylus file](https://github.com/Mister-Hope/vuepress-theme-hope/blob/master/styles/nightMode.styl)? I think it’s good enough.
You can just import it in index.styl and add a “theme-night” class to <body> tag when dark mode is on.

**Comment by Mister-Hope** [comment]
I make a plugin [here](https://github.com/Mister-Hope/vuepress-theme-hope/tree/master/packages/theme-color#readme) You can check this out. If it is okey, I will make a PR

**Comment by superusercode** [comment]
any news on this issue? assuming the offer from Mister-Hope still stands

**Comment by quannetwork** [comment]
Love this feature <3. Still looking forward to it.

**Comment by Mister-Hope** [comment]
@kefranabg I can open a PR for this. If you need it, just call me.

**Comment by Mister-Hope** [comment]
@kefranabg  check this https://github.com/vuejs/vuepress/pull/2232

**Comment by KieranHunt** [comment]
I recently wanted to add dark mode to our own vuepress site, did a little survey around what was available, and wanted to report my findings:

- [vuepress-theme-default-prefers-color-scheme](https://github.com/tolking/vuepress-theme-default-prefers-color-scheme) by @tolking does basically what I'm looking for.
  - I really like the `prefers-color-scheme` approach.
  - I just find the light mode code blocks to be a bit low-contrast
  - I'd ideally like it if it were supported in the default theme.
- @Mister-Hope's [pull request](https://github.com/vuejs/vuepress/pull/2232) is cool.
  - I do find it a little too dark (I don't think dark mode should be pitch black)
  - It doesn't seem to use `prefers-color-scheme`.
  - I swear I've seen that light/dark toggle around before.  Does it come from somewhere else?

**Comment by Mister-Hope** [comment]
> * @Mister-Hope's [pull request](https://github.com/vuejs/vuepress/pull/2232) is cool.
>   
>   * I do find it a little too dark (I don't think dark mode should be pitch black)

Maybe you are right, but all my device are amoled, and I work in pure dark. So a real dark fits me at least

>   * It doesn't seem to use `prefers-color-scheme`.

It's modified from `@mr-hope/vuepress-plugin-theme-color`, which is built for my theme [vuepress-theme-hope](https://github.com/Mister-Hope/vuepress-theme-hope)

>   * I swear I've seen that light/dark toggle around before.  Does it come from somewhere else?

Yes, I got it from a switch collection with license MIT, but I am not sure if the collection owner built it. It's a nice remind, I will try to check it's license, if I cannot find it, I will change a new one.

**Comment by Mister-Hope** [comment]
> I just find the light mode code blocks to be a bit low-contrast

I am thinking that too, but `prism` 's other theme usually have worse apprence. And I am not modifying it, it's color just remains which the theme-default provides.

**Comment by tolking** [comment]
>   * I really like the `prefers-color-scheme` approach.

I really like the `prefers-color-scheme` approach. Switch light and dark modes according to system settings.

**Comment by Mister-Hope** [comment]
@KieranHunt I change the apprence and the switch with some new commits.

I had a look at `prefers-color-scheme`. My first idea is to make a plugin, so I just cover some of the styles with higher order css selectors.

We can switch to `prefers-color-scheme` actually, but that also means I have to modify styles in `@vuepress/theme-default` heavily, so,

>To avoid wasting your time, it's best to open a feature request issue first and wait for approval before working on it.

I shall first ask the team. @kefranabg  @meteorlxy @vinayakkulkarni

**Comment by BboyAkers** [comment]
Is it possible to make it toggle-able for the end user?

**Comment by god-appel** [comment]
Some very interesting take-aways from this writeup by Stackoverlow which we need to keep in mind: https://stackoverflow.blog/2020/03/31/building-dark-mode-on-stack-overflow/

And with that - is there an update on this?

**Comment by Dorson** [comment]
Made a flexible solution for the Dark Mode or any CSS color theme:

https://dorson.github.io/CSS-Dark-Mode-and-color-switch/

Uses simple JS functions and CSS variables to set colors dynamically.

Have fun !

**Comment by Mister-Hope** [comment]
I rebuild my theme with `prefers-color-scheme` @KieranHunt , you can have a look at this.

I will ship this feature to the `@vuepress/theme-default` as a new feature instead of a plugin in a new PR.

**Comment by Mister-Hope** [comment]
I have opened a new PR #2301.

This new darkmode is using `prefers-color-scheme` with css vars, and will fall back to "6 - 18 lightmode; 18 - 6 darkmode" if the user's broswer does not support `prefers-color-scheme`.

Also, the code block apprearence in lightmode have be rebuilt.

Any suggestions are welcomed. 😃

**Comment by Mister-Hope** [comment]
I was wondering if I should put a switch with "light / auto / default" somewhere in the theme, and if you can give some opinions, it would be wonderful.

Besides, if you want to add this switch, please tell me where do you want to put it, becauze I have no idea about it.

**Comment by Dorson** [comment]
> I was wondering if I should put a switch with "light / auto / default" somewhere in the theme, and if you can give some opinions, it would be wonderful.
> 
> Besides, if you want to add this switch, please tell me where do you want to put it, because I have no idea about it.

Yes, such a toggle switch would be perfect in sunny conditions. Could be as minimal as possible, because most of the time we will not use it as the users. Maybe 3 times per day. In the sunshine on mobile screens, or when it's evening time.

White Unicode emoticon character with a dark background would work.  `🌞 or 🌅 or 🌙 or 🌒 or ❄️ (for white only) or ◑ or 💡 or ⛯  `

Such a night / dark mode toggle switch could be placed near the top menu icon, or the search bar. Top right corner, or left right corner, or the bottom left corner. The bottom right corner is assumed to be used for the scroll up buttons, because the scroll bars end in that screen location.

Some websites place it on the lower left side. Most websites place it near the top right corner, or near the menu icon.

**Comment by Mister-Hope** [comment]
I put it in the navbar in the last PR, while I found it too narrow if I placed it there. Fixed it in a corner can be a good choice. i will try later. Thanks.

**Comment by Dorson** [comment]
> I put it in the navbar in the last PR, while I found it too narrow if I placed it there. Fixed it in a corner can be a good choice. i will try later. Thanks.

For the demo page, I put the Night / Light switch in the top right corner : 
https://dorson.github.io/CSS-Dark-Mode-and-color-switch/

It all depends on the other elements of the page and the text flow direction. Arabic and Hebrew is  like a mirror position to the left to right text.

**Comment by quannetwork** [comment]
:100: @Dorson your repo is great. Good fit for my Jekyll themes.

**Comment by Mister-Hope** [comment]
[Here is a preview](https://deploy-preview-2301--vuepress.netlify.com/)

A RFC here, what do you think about the switch? 😃

**Comment by KieranHunt** [comment]
That's very cool, @Mister-Hope

**Comment by yonkov** [comment]
Hey guys, I have written a tutorial how to implement night mode switch in the default theme. Check this tutorial: 
[Switching off the Lights Part Three - Adding Dark Mode to VuePress](https://yonkov.github.io/post/add-dark-mode-to-vuepress/)

However, it will be also nice to be able to do that in the top menu. Currently the defaul theme does not support compoments as menu items I think?

**Comment by Mister-Hope** [comment]
> @Mister-Hope, good job!  I also made a small contribution to this topic by writing a tutorial how to implement night mode switch in the default vuepress theme. Check this article: 
> [Switching off the Lights Part Three - Adding Dark Mode to VuePress](https://yonkov.github.io/post/add-dark-mode-to-vuepress/)
> 
> Basically, I am creating a vue component and injecting it inside the markdown file of the home page. However, it will be even better to be able to add the dark mode toggle in the top menu. Currently the default theme does not support components as menu items I think?

Yes, but you can change it by overiding Navbar.vue. On the size around 700-800px the navbar does not have a good appearance because of too many items. So I did not place it in navbar.

**Comment by DMiradakis** [comment]
This is all **_incredible_**. Thank you, everyone! I love VuePress.

**Comment by shindodkar** [comment]
is this open and ready to make changes ? I would like to contribute on this

**Comment by Mister-Hope** [comment]
We are moving on to vuepress2 now, and this project is no longer maintained (At least I have a dozen of PRs unmerged)
